README.md
# NAME

<div>
    <a href="https://travis-ci.org/PACKED-vzw/Catmandu-Store-Resolver"><img src="https://travis-ci.org/PACKED-vzw/Catmandu-Store-Resolver.svg?branch=master"></a>
</div>

Catmandu::Resolver - Store/retrieve items from the Resolver

# SYNOPSIS

    # From the command line
    $ catmandu export Resolver --id 1234 --url http://www.resolver.be --username demo --password demo to YAML
    ---
    data:
        data_pids:
            - https://resolver.be/collection/work/data/9031
        documents:
            - 88682
            - 88683
        domain: https://resolver.be
        id: '9031'
        persistentURIs:
            - https://resolver.be/collection/9031
            - https://resolver.be/collection/9031/untitled
            - https://resolver.be/collection/work/data/9031/html
            - https://resolver.be/collection/work/data/9031/html/untitled
            - https://resolver.be/collection/work/data/9031
            - https://resolver.be/collection/work/representation/9031/1
            - https://resolver.be/collection/work/representation/9031/1/untitled
            - https://resolver.be/collection/work/representation/9031
        type: work
        work_pid: https://resolver.be/collection/9031

    ...

    # From a Catmandu Fix
    lookup_in_store(
        objectNumber,
        Resolver,
        username: username,
        password: password,
        url: http://www.resolver.be
    )

    # Create or retrieve a PID from a fix
    make_pid(
        path,
        'http://www.resolver.be',
        username,
        password,
        -type: work
    )

# DESCRIPTION

Configure the [Resolver](https://github.com/PACKED-vzw/resolver) as a [store](http://librecat.org/Catmandu/#stores) for [Catmandu](http://librecat.org/).

Museum objects and records require a PID to be uniquely identifiable. The Resolver tool
generates and resolves these PIDs. By using this store, PIDs can be queried (based on
the object number of the record as stored in the resolver), created, updated and deleted
from Catmandu.

# MODULES

- [Catmandu::Store::Resolver](https://metacpan.org/pod/Catmandu::Store::Resolver)
- [Catmandu::Fix::make\_pid](https://metacpan.org/pod/Catmandu::Fix::make_pid)

# SEE ALSO

[Catmandu](https://metacpan.org/pod/Catmandu)

# AUTHORS

Pieter De Praetere, `pieter at packed.be`

# CONTRIBUTORS

Pieter De Praetere, `pieter at packed.be`

# COPYRIGHT AND LICENSE

This package is copyright (c) 2016 by PACKED vzw.
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Catmandu-Store-Resolver

{{$NEXT}}

0.06  2017-02-20 13:45:49 UTC
        - Added Catmandu::Fix::make_pid
        - Renamed to Catmandu-Resolver

0.05  2017-02-17 14:46:52 UTC
        - Added PID as _id in the return value of the store

0.04  2017-01-11 09:50:43 UTC
	- Bugfix: don't die on 404

0.03  2017-01-10 12:33:51 UTC
        - Bugfix: 403 error on multiple submits
        - Updated documentation

0.02  2017-01-04 13:11:46 UTC
	- Added support for PUT/POST/DELETE

0.01  2017-01-03 11:15:45 UTC
        - Initial release


lib/Catmandu/Store/Resolver/API.pm
package Catmandu::Store::Resolver::API;

use Moo;
use JSON;
use LWP::UserAgent;

use Catmandu::Sane;

has url      => (is => 'ro', required => 1);
has username => (is => 'ro', required => 1);
has password => (is => 'ro', required => 1);

has ua         => (is => 'lazy');
has cookie_jar => (is => 'lazy');

sub _build_ua {
    my $self = shift;
    my $ua = LWP::UserAgent->new();
    $ua->cookie_jar({});
    return $ua;
}

sub _build_cookie_jar {
    my $self = shift;
    return $self->login();
}

sub login {
    my $self = shift;
    my $auth_url = '%s/resolver/api/login';
    my $req_url = sprintf($auth_url, $self->url);
    my $form = {
        'username' => $self->username,
        'password' => $self->password
    };
    my $response = $self->ua->post($req_url, Content => $form);
    if ($response->is_success) {
        return $self->ua->cookie_jar;
    } else {
        Catmandu::HTTPError->throw({
            code             => $response->code,
            message          => $response->status_line,
            url              => $response->request->uri,
            method           => $response->request->method,
            request_headers  => [],
            request_body     => $response->request->decoded_content,
            response_headers => [],
            response_body    => $response->decoded_content
        });
        return undef;
    }
}

sub logout {
    my $self = shift;
    my $url = '%s/resolver/api/logout';
    my $req_url = sprintf($url, $self->url);
    $self->ua->cookie_jar($self->cookie_jar);

    my $response = $self->ua->get($req_url);
    if ($response->is_success) {
        return 1;
    } else {
        Catmandu::HTTPError->throw({
            code             => $response->code,
            message          => $response->status_line,
            url              => $response->request->uri,
            method           => $response->request->method,
            request_headers  => [],
            request_body     => $response->request->decoded_content,
            response_headers => [],
            response_body    => $response->decoded_content
        });
        return undef;
    }
}

sub get {
    my ($self, $id) = @_;
    my $url = '%s/resolver/api/entity/original/%s';
    my $req_url = sprintf($url, $self->url, $id);

    $self->ua->cookie_jar($self->cookie_jar);

    my $response = $self->ua->get($req_url);

    if ($response->is_success) {
        # New API
        return decode_json($response->decoded_content);
    } else {
        # Try with the old API
        # Please note: this doesn't work with the original object number
        # so it could say 'does not exist' because the original object number
        # is not equal to the PID
        $url = '%s/resolver/api/entity/%s';
        $req_url = sprintf($url, $self->url, $id);
        $response = $self->ua->get($req_url);
        if ($response->is_success) {
            return decode_json($response->decoded_content);
        } else {
            # Give up
            Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content
            });
            return undef;
        }
    }
}

sub post {
    my ($self, $data) = @_;
    my $json_data = encode_json($data);
    my $url = '%s/resolver/api/entity';
    my $req_url = sprintf($url, $self->url);

    $self->ua->cookie_jar($self->login());

    my $response = $self->ua->post($req_url, Content_Type => 'application/json', Content => $json_data);

    if ($response->is_success) {
        return decode_json($response->decoded_content);
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content
            });
        return undef;
    }

}

sub put {
    my ($self, $id, $data) = @_;
    my $json_data = encode_json($data);
    my $url = '%s/resolver/api/entity/%s';
    my $req_url = sprintf($url, $self->url, $id);

    $self->ua->cookie_jar($self->login());

    my $response = $self->ua->put($req_url, Content_Type => 'application/json', Content => $json_data);

    if ($response->is_success) {
        return decode_json($response->decoded_content);
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content
            });
        return undef;
    }
}

sub delete {
    my ($self, $id) = @_;
    my $url = '%s/resolver/api/entity/%s';
    my $req_url = sprintf($url, $self->url, $id);

    $self->ua->cookie_jar($self->login());

    my $response = $self->ua->delete($req_url);

    if ($response->is_success) {
        return $response->decoded_content;
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content
            });
        return undef;
    }
}

1;

lib/Catmandu/Store/Resolver/Bag.pm
package Catmandu::Store::Resolver::Bag;

use Moo;
use JSON;

use Catmandu::Sane;
use Catmandu::Store::Resolver::API;

with 'Catmandu::Bag';

has api => (is => 'lazy');

sub _build_api {
    my $self = shift;
    my $api = Catmandu::Store::Resolver::API->new(
        url => $self->store->url,
        username => $self->store->username,
        password => $self->store->password
    );
    return $api;
}

sub generator {
    my $self = shift;
}

sub get {
    my ($self, $id) = @_;
    return $self->api->get($id);
}

sub add {
    my ($self, $data) = @_;
    return $self->api->post($data);
}

sub update {
    my ($self, $id, $data) = @_;
    return $self->api->put($id, $data);
}

sub delete {
    my ($self, $id) = @_;
    return $self->api->delete($id);
}

sub delete_all {
    my $self = shift;
}

1;

lib/Catmandu/Resolver.pm
package Catmandu::Resolver;

our $VERSION = '0.05';

use strict;
use warnings;

use Moo;
use Catmandu::Sane;

1;
__END__

=head1 NAME

=for html <a href="https://travis-ci.org/PACKED-vzw/Catmandu-Store-Resolver"><img src="https://travis-ci.org/PACKED-vzw/Catmandu-Store-Resolver.svg?branch=master"></a>

Catmandu::Resolver - Store/retrieve items from the Resolver

=head1 SYNOPSIS

    # From the command line
    $ catmandu export Resolver --id 1234 --url http://www.resolver.be --username demo --password demo to YAML
    ---
    data:
        data_pids:
            - https://resolver.be/collection/work/data/9031
        documents:
            - 88682
            - 88683
        domain: https://resolver.be
        id: '9031'
        persistentURIs:
            - https://resolver.be/collection/9031
            - https://resolver.be/collection/9031/untitled
            - https://resolver.be/collection/work/data/9031/html
            - https://resolver.be/collection/work/data/9031/html/untitled
            - https://resolver.be/collection/work/data/9031
            - https://resolver.be/collection/work/representation/9031/1
            - https://resolver.be/collection/work/representation/9031/1/untitled
            - https://resolver.be/collection/work/representation/9031
        type: work
        work_pid: https://resolver.be/collection/9031

    ...

    # From a Catmandu Fix
    lookup_in_store(
        objectNumber,
        Resolver,
        username: username,
        password: password,
        url: http://www.resolver.be
    )

    # Create or retrieve a PID from a fix
    make_pid(
        path,
        'http://www.resolver.be',
        username,
        password,
        -type: work
    )

=head1 DESCRIPTION

Configure the L<Resolver|https://github.com/PACKED-vzw/resolver> as a L<store|http://librecat.org/Catmandu/#stores> for L<Catmandu|http://librecat.org/>.

Museum objects and records require a PID to be uniquely identifiable. The Resolver tool
generates and resolves these PIDs. By using this store, PIDs can be queried (based on
the object number of the record as stored in the resolver), created, updated and deleted
from Catmandu.

=head1 MODULES

=over

=item L<Catmandu::Store::Resolver>

=item L<Catmandu::Fix::make_pid>

=back

=head1 SEE ALSO

L<Catmandu>

=head1 AUTHORS

Pieter De Praetere, C<< pieter at packed.be >>

=head1 CONTRIBUTORS

Pieter De Praetere, C<< pieter at packed.be >>

=head1 COPYRIGHT AND LICENSE

This package is copyright (c) 2016 by PACKED vzw.
This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut